### PR TITLE
Issue 226  broadcast op

### PR DESCRIPTION
**Pull request by adler-j**
Added broadcast and reduction operators

Log is somewhat convoluted due to a merge from master. Needs to be done anyway I guess.

### CODE REVIEW

File: odl/operator/pspace_ops.py [hunk 1]
```diff
@@ -562,6 +606,249 @@ def adjoint(self):
         return ComponentProjection(self.range, self.index)
 
 
+class BroadcastOperator(Operator):
```
Comment (adler-j) [inline odl/operator/pspace_ops.py]: This is the important part to review.

File: odl/operator/pspace_ops.py [hunk 2]
```diff
@@ -562,6 +606,249 @@ def adjoint(self):
...
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op.adjoint([1, 2, 3])
+        ProductSpace(Rn(3), 2).element([
+            [1.0, 2.0, 3.0],
+            [2.0, 4.0, 6.0]
+        ])
+        """
+        return BroadcastOperator(*[op.adjoint for op in self.operators])
+
+
```
Comment (adler-j) [inline odl/operator/pspace_ops.py]: Review ends here.

File: odl/operator/pspace_ops.py [hunk 3]
```diff
@@ -562,6 +606,273 @@ def adjoint(self):
...
+        >>> I = odl.IdentityOperator(odl.Rn(3))
+        >>> op = BroadcastOperator(I, 2 * I)
+        >>> op.adjoint([[1, 2, 3], [2, 3, 4]])
+        Rn(3).element([5.0, 8.0, 11.0])
+        """
+        return ReductionOperator(*[op.adjoint for op in self.operators])
+
+
+class ReductionOperator(Operator):
+    """Reduce argument to over set of operators.
```
Comment (adler-j) [inline odl/operator/pspace_ops.py]: remove to

File: odl/operator/pspace_ops.py [hunk 4]
```diff
@@ -562,6 +606,273 @@ def adjoint(self):
         return ComponentProjection(self.range, self.index)
 
 
+class BroadcastOperator(Operator):
+    """Broadcast argument to set of operators.
+
+    An argument is broadcasted by evaluating several operators in the same
```
Comment (kohr-h) [inline odl/operator/pspace_ops.py]: broadcast (without "ed")

File: odl/operator/pspace_ops.py [hunk 5]
```diff
@@ -562,6 +606,273 @@ def adjoint(self):
...
+        >>> op(x)
+        ProductSpace(Rn(3), 2).element([
+            [1.0, 2.0, 3.0],
+            [2.0, 4.0, 6.0]
+        ])
+        >>> op.derivative(x)(x)
+        ProductSpace(Rn(3), 2).element([
+            [1.0, 2.0, 3.0],
+            [2.0, 4.0, 6.0]
+        ])
```
Comment (kohr-h) [inline odl/operator/pspace_ops.py]: Can you find an example where the derivative is different from the original operator? Maybe use `ResidualOp`, (in one component), should be simple enough.

File: odl/operator/pspace_ops.py [hunk 6]
```diff
@@ -562,6 +606,273 @@ def adjoint(self):
...
+        """A tuple of sub-operators"""
+        return self._operators
+
+    def _call(self, x, out=None):
+        """Apply operators to ``x`` and sum.
+
+        Parameters
+        ----------
+        x : domain element
+            Input vector to be extended
```
Comment (kohr-h) [inline odl/operator/pspace_ops.py]: copy-paste

File: odl/operator/pspace_ops.py (hunk 2 above)
Comment (kohr-h) [inline odl/operator/pspace_ops.py]: All good I'd say, apart from some doc stuff. The weighting issue lurks also here, but an upstream fix to that will fix this, too.

File: odl/operator/pspace_ops.py (hunk 5 above)
Comment (adler-j) [inline odl/operator/pspace_ops.py]: Good example, I'll do that.

**Pull request merged**